metadata-ingestion/src/datahub/ingestion/source/feast.py

```diff
@@ -0,0 +1,266 @@
+                            sources=entity_sources,
+                        )
+                    )
+
+                    # make the MCE and workunit
+                    mce = MetadataChangeEvent(proposedSnapshot=entity_snapshot)
+                    wu = MetadataWorkUnit(id=entity["name"], mce=mce)
+                    self.report.report_workunit(wu)
+                    yield wu
+
+                # ingest features in table
+                for feature in table["features"]:
+
+                    # create snapshot instance for the feature
+                    feature_snapshot = MLFeatureSnapshot(
+                        urn=builder.make_ml_feature_urn(table["name"], feature["name"]),
+                        aspects=[],
+                    )
+
+                    feature_sources = []
+
+                    if feature["batch_source"] is not None:
+                        feature_sources.append(
+                            builder.make_dataset_urn(
+                                feature["batch_source_platform"],
+                                feature["batch_source_name"],
+                                self.config.env,
+                            )
+                        )
+
+                    if feature["stream_source"] is not None:
+                        feature_sources.append(
+                            builder.make_dataset_urn(
+                                feature["stream_source_platform"],
+                                feature["stream_source_name"],
+                                self.config.env,
+                            )
+                        )
+
+                    # append feature name and type
+                    feature_snapshot.aspects.append(
+                        MLFeaturePropertiesClass(
+                            dataType=self.get_field_type(
+                                feature["type"], feature["name"]
+                            ),
+                            sources=feature_sources,
+                        )
+                    )
+
+                    # make the MCE and workunit
+                    mce = MetadataChangeEvent(proposedSnapshot=feature_snapshot)
+                    wu = MetadataWorkUnit(id=feature["name"], mce=mce)
+                    self.report.report_workunit(wu)
+                    yield wu
+
+                featuretable_snapshot = MLFeatureTableSnapshot(
+                    urn=builder.make_ml_feature_table_urn(platform, table["name"]),
+                    aspects=[],
+                )
+
+                featuretable_snapshot.aspects.append(
+                    MLFeatureTablePropertiesClass(
+                        mlFeatures=[
+                            builder.make_ml_feature_urn(
+                                table["name"],
+                                feature["name"],
+                            )
+                            for feature in table["features"]
+                        ],
+                        mlPrimaryKeys=[
```

Comment:
there can be multiple primary keys for a feature table?

Reply:
Yes – they act as a composite key

Reply:
got it... let's make sure we document that in any field comments